# ML-03 — Frame Your Lane as an ML Task

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/samhere3116-maker/ML-Pipeline/blob/main/work/notebooks/w02_ml_task_framing.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

This is a classification problem — I'm trying to sort pages into "declining" or "not declining." But the actual output I want isn't just a yes/no, it's a ranked list, so I use the model's probability score to sort pages from most-likely-to-need-review to least.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

I'm using trend_direction == "down" as my target for now. It's not a perfect target though — it's just a label based on where things stand right now, not whether the page will actually keep declining. A better version later would predict what happens in the next 30 days based on the last 90 days, instead of just labeling the current snapshot.

## 3. Success metric

*One metric you can defend. What number means 'good'?*

I'm using Precision@50. The reason is simple — a reviewer can only realistically check about 50 pages, not all 30,000. So what matters isn't "is the model right overall," it's "are the top 50 pages it picks actually worth checking." Regular accuracy wouldn't mean much here since most pages are boring and easy to get right by default.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

url = "https://raw.githubusercontent.com/samhere3116-maker/ML-Pipeline/main/data/raw/content_refresh_anonymized.csv"
df = pd.read_csv(url)
df = df[(df["impressions_90d"] > 0) & (df["content_age_days"] >= 90)].drop_duplicates(subset="content_id")

print(f"Shape: {df.shape[0]} rows, {df.shape[1]} columns")
df[["content_id", "impressions_90d", "avg_position", "content_age_days", "trend_direction"]].head()




Shape: 30000 rows, 44 columns


,content_id,impressions_90d,avg_position,content_age_days,trend_direction
0,content_304f48230142,3803,10.6,187,down
1,content_a1fb4e703a9e,15320,20.3,445,down
2,content_9aa793d4d895,12581,36.5,141,down
3,content_331d6c4de07b,11751,6.2,463,stable
4,content_d99b7a2d90ca,19140,44.0,263,down


## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

In [2]:
baseline_precision_at_50 = 0.240
model_precision_at_50 = 0.740
print(f"Baseline rule: {baseline_precision_at_50}  |  Learned model: {model_precision_at_50}")

Baseline rule: 0.24  |  Learned model: 0.74


A simple rule breaks down because "declining" depends on more than one thing at once. A stale page with barely any traffic was never worth reviewing anyway, but a stale page that used to get a lot of traffic is a real problem — same "staleness," totally different situation depending on the traffic. A basic if-statement can't easily juggle that combo. I already tested this myself: my hand-written rule only got 24% right on its top 50 picks, while a simple decision tree got 82% — proof the real pattern isn't a single clean threshold.



## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.